### PR TITLE
[ENG-7300] Upgrade Node Component Reordering to V2.

### PR DESCRIPTION
**Pull request by Ostap-Zherebetskyi**
## Purpose

Upgrade Node Component Reordering to V2.

## Changes

Added API endpoint to reorder components
Updated node children endpoint for correct component ordering

## QA Notes

N/A

## Documentation

N/A

## Side Effects

N/A

## Ticket

https://openscience.atlassian.net/browse/ENG-7300

### CODE REVIEW

File: api/base/views.py [hunk 1]
```diff
@@ -485,7 +489,10 @@ def get_queryset(self):
         auth = get_user_auth(self.request)
         node_pks = node.node_relations.filter(is_node_link=False).select_related('child')\
             .values_list('child__pk', flat=True)
-        return self.get_queryset_from_request().filter(pk__in=node_pks).can_view(auth.user, auth.private_link).order_by('-modified')
+        order = '-modified'
```
Comment (Johnetordoff) [inline api/base/views.py:492]: Similar to above don't hard code default values here, get them from the view.

File: api/nodes/views.py [hunk 2]
```diff
@@ -2358,3 +2359,83 @@ def get_default_queryset(self):
...
+                node_relation_obj = node_relations.filter(child_id=child_node_id)
+                if node_relation_obj.exists():
+                    node_relation_id = node_relation_obj.first().id
+                    new_node_relation_ids.remove(node_relation_id)
+                    new_node_relation_ids.insert(node_pos['_order'], node_relation_id)
+            except NotFound:
+                errors.append(node_pos.get('id'))
+
+        if errors:
+            return JsonResponse(
```
Comment (Johnetordoff) [inline api/nodes/views.py:2434]: Probably should use a `ValidationError` from `from rest_framework.exceptions import ValidationError` instead of `JsonResponse`

File: api_tests/nodes/views/test_node_children_list.py [hunk 3]
```diff
@@ -301,6 +301,39 @@ def test_node_child_filtering(self, app, user):
...
+
+    def test_node_child_ordering(self, app, user):
+        project = ProjectFactory(creator=user)
+
+        title_one, title_two = fake.bs(), fake.bs()
+        NodeFactory(title=title_one, parent=project)
+        NodeFactory(title=title_two, parent=project)
+        component_order = [el.child._id for el in project.node_relations.all()]
+
+        url = '/{}nodes/{}/children/?_order=component_order'.format(
```
Comment (Johnetordoff) [inline api_tests/nodes/views/test_node_children_list.py:315]: try to use f-strings

File: api_tests/nodes/views/test_node_reorder_components.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    def test_reorder_components_write_read(
+            self, app, write_contrib, read_contrib, url, payload_asc
+    ):
+        res = app.patch_json_api(url, payload_asc, auth=write_contrib.auth, expect_errors=True, bulk=True)
+        assert res.status_code == 403
+
+        res = app.patch_json_api(url, payload_asc, auth=read_contrib.auth, expect_errors=True, bulk=True)
+        assert res.status_code == 403
+
+    def test_reorder_invalid_component(
```
Comment (Johnetordoff) [inline api_tests/nodes/views/test_node_reorder_components.py:113]: There should be more test cases in case of invalid payload and conflicting instructions, including the following:

- _order is higher the list is long (for one and multiple payload items)
- _order is lower then zero long (for one and multiple payload items)
- two or more items in the payload have the same _order
- an item id appears in the payload >1 time with the same _order information
 - an item id appears in the payload >1 time with the different _order information

That a summary of the types of validation cases I'd like to have descriptive error messages for, it doesn't need a unique message for every one, but it should distinguish 400s Invalid Payload from 409 Conflicting Information and let the develop know the nature of the mistake they've made.

File: api/base/views.py [hunk 5]
```diff
@@ -485,7 +489,10 @@ def get_queryset(self):
...
+        if self.request.query_params.get('_order', None) == 'component_order':
```
Comment (Johnetordoff) [inline api/base/views.py:493]: I think this is making it a little more difficult than it has to be. Ideally instead of `/?_order=component_order` you should use DRF existing `sort` param and `/?sort=_order` . I would look and `OSFOrderingFilter` and potentially add a `_order` field to the serializer.

**Review by Johnetordoff — CHANGES_REQUESTED** [review]
There's some issues with the params sorting for the GET functionality and the need for more tests, but a good first effort, 👍

**Review by Johnetordoff — APPROVED** [review]
Good work tying down all those edge cases, thanks.
